Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrMetadataEmitter.kt

```diff
@@ -0,0 +1,333 @@
+            } else {
+                KmAnnotationArgument.ULongValue(value)
+            }
+
+            else -> error("Integral constant of value $value with unexpected size of $size.")
+        }
+        is DoubleConstantStub -> when (size) {
+            4 -> KmAnnotationArgument.FloatValue(value.toFloat())
+            8 -> KmAnnotationArgument.DoubleValue(value)
+            else -> error("Floating-point constant of value $value with unexpected size of $size.")
+        }
+    }
+
+    private val TypeParameterType.id: Int
+        get() = TODO()
+
+    private val TypeParameterStub.id: Int
+        get() = TODO()
+
+    private val FunctionParameterStub.flags: Flags
+        get() = flagsOf()
+}
+
+private class StubIrUniqIdProvider {
+    private val mangler = KotlinLikeInteropMangler()
+
+    fun uniqIdForFunction(function: FunctionStub): UniqId = with(mangler) {
+        when (function.origin) {
+            is StubOrigin.Function -> function.origin.function.uniqueSymbolName
+            is StubOrigin.ObjCMethod -> {
+                require(this.context is ManglingContext.Entity) {
+                    "Unexpected mangling context $context for method ${function.name}."
+                }
+                function.origin.method.uniqueSymbolName
+            }
+            // TODO: What to do with "create" method in Objective-C categories?
+            else -> error("Unexpected origin ${function.origin} for function ${function.name}.")
+        }.toUniqId()
+    }
+
+    fun uniqIdForProperty(property: PropertyStub): UniqId = with (mangler) {
+        when (property.origin) {
+            is StubOrigin.ObjCProperty -> {
+                require(this.context is ManglingContext.Entity) {
+                    "Unexpected mangling context $context for property ${property.name}."
+                }
+                property.origin.property.uniqueSymbolName
+            }
+            is StubOrigin.Constant -> property.origin.constantDef.uniqueSymbolName
+            is StubOrigin.Global -> property.origin.global.uniqueSymbolName
+            // TODO: What to do with origin for enum entries and struct fields?
+            else -> error("Unexpected origin ${property.origin} for property ${property.name}.")
+        }.toUniqId()
+    }
+
+    fun uniqIdForTypeAlias(typeAlias: TypealiasStub): UniqId = with (mangler) {
+        when (typeAlias.origin) {
+            is StubOrigin.TypeDef -> typeAlias.origin.typedefDef.uniqueSymbolName
+            is StubOrigin.Enum -> typeAlias.origin.enum.uniqueSymbolName
+            else -> error("Unexpected origin ${typeAlias.origin} for typealias ${typeAlias.alias.fqName}.")
+        }
+    }.toUniqId()
+
+    private fun String.toUniqId() = UniqId(cityHash64())
+}
+
+/**
+ * Translates single [StubContainer] to [KmModuleFragment].
+ */
+internal class ModuleMetadataEmitter(
+        private val packageFqName: String
+) : StubIrToMetadataTranslationExtensions, StubIrVisitor<StubContainer?, Unit> {
```

Comment:
> `StubIrToMetadataTranslationExtensions`

Isn't this an abuse of inheritance?

Reply:
Yeah, definitely. I tried to minimize scope of these functions, but it really looks ugly. Replaced with private top-level declarations instead.